### PR TITLE
fix(iroh-net): fix memory leaks in the iroh-relay server

### PR DESCRIPTION
**Pull request by dignifiedquire**
We observed continous memory increase on the relays, this fixes the main offenders we have found. Mainly the issue was not cleaning up tasks that were spawned into `JoinSet`s.

### CODE REVIEW

**Review by Arqu — APPROVED** [review]
A bunch of testing and more profiling later, can confirm this seems to fix the issue. Prod is the real test :P

File: iroh-net/src/relay/server/http_server.rs [hunk 1]
```diff
@@ -542,10 +550,12 @@ impl RelayService {
...
     {
-        hyper::server::conn::http1::Builder::new()
+        let conn = hyper::server::conn::http1::Builder::new()
             .serve_connection(hyper_util::rt::TokioIo::new(io), self)
-            .with_upgrades()
-            .await?;
+            .with_upgrades();
+
+        conn.await?;
+
```
Comment (matheus23) [inline iroh-net/src/relay/server/http_server.rs:558]: A little bit of diff noise

File: iroh-net/src/relay/server/http_server.rs [hunk 2]
```diff
@@ -253,11 +253,19 @@ impl ServerState {
...
+                                panic!("task panicked: {:#?}", err);
+                            }
+                        }
+                    }
                     res = listener.accept() => match res {
                         Ok((stream, peer_addr)) => {
                             debug!("[{http_str}] relay: Connection opened from {peer_addr}");
                             let tls_config = tls_config.clone();
                             let service = service.clone();
+
```
Comment (matheus23) [inline iroh-net/src/relay/server/http_server.rs:268]: Diff noise

**Review by matheus23 — APPROVED** [review]
I'm excited by this PR (& the idea of a 8.5MB memory-at-idle relay).

Usually, I wouldn't care much about diff noise, but given there will likely be conflicts with #2873, we should reduce diff noise as much as possible.

**Review by rklaehn — APPROVED** [review]

**Review by flub — APPROVED** [review]

File: iroh-net/src/relay/server/http_server.rs (hunk 1 above)
Comment (dignifiedquire) [reply iroh-net/src/relay/server/http_server.rs:558]: fixed

File: iroh-net/src/relay/server/http_server.rs (hunk 2 above)
Comment (dignifiedquire) [reply iroh-net/src/relay/server/http_server.rs]: fixed

**Pull request merged**